lib/engine/game/g_1858/game.rb

```diff
@@ -46,6 +48,145 @@ class Game < Game::Base
+              tile = hex.tile
+              next unless tile.color == :white
+
+              hex.neighbors.each do |edge, neighbor|
+                next unless home_hexes.include?(neighbor)
+
+                stub = Engine::Part::Stub.new(edge)
+                tile.stubs << stub
+                stubs[minor] << { tile: tile, stub: stub }
+              end
+            end
+          end
+
+          stubs
+        end
+
+        # Removes the stubs from the private railway's home hexes.
+        def release_stubs(minor)
+          stubs = @stubs[minor]
+          return unless stubs
+
+          stubs.each { |tile_stub| tile_stub[:tile].stubs.delete(tile_stub[:stub]) }
+          @stubs.delete(minor)
+        end
+
+        def clear_graph_for_entity(_entity)
+          @graph.clear
+          @graph_broad.clear
+          @graph_metre.clear
+        end
+
+        def clear_token_graph_for_entity(entity)
+          clear_graph_for_entity(entity)
+        end
+
+        def operating_round(round_num = 1)
+          @round_num = round_num
+          Engine::Round::Operating.new(self, [
+            G1858::Step::Track,
+            Engine::Step::Token,
+            Engine::Step::Route,
+            Engine::Step::Dividend,
+            Engine::Step::DiscardTrain,
+            Engine::Step::BuyTrain,
+            Engine::Step::IssueShares,
+          ], round_num: round_num)
+        end
+
+        # Returns the company object for a private railway given its associated
+        # minor object. If passed a company then returns that company.
+        def private_company(entity)
+          return entity if entity.company?
+
+          @companies.find { |company| company.sym == entity.id }
+        end
+
+        # Returns the minor object for a private railway given its associated
+        # company object. If passed a minor then returns that minor.
+        def private_minor(entity)
+          return entity if entity.minor?
+
+          @minors.find { |minor| minor.id == entity.sym }
+        end
+
+        # Returns true if the hex is this private railway's home hex.
+        def home_hex?(operator, hex)
+          operator.coordinates.include?(hex.coordinates)
+        end
+
+        def tile_lays(entity)
+          entity.corporation? ? TILE_LAYS : MINOR_TILE_LAYS
+        end
+
+        def metre_gauge_upgrade(old_tile, new_tile)
+          # Check if the only new track on the tile is metre gauge
+          old_track = old_tile.paths.map(&:track)
+          new_track = new_tile.paths.map(&:track)
+          old_track.each { |t| new_track.slice!(new_track.index(t) || new_track.size) }
+          new_track.uniq == [:narrow]
+        end
+
+        def upgrade_cost(tile, hex, entity, _spender)
+          return 0 if tile.upgrades.empty?
+          return 0 if entity.minor? && home_hex?(entity, hex)
+
+          cost = tile.upgrades[0].cost
+          return cost unless metre_gauge_upgrade(tile, hex.tile)
+
+          discount = cost / 2
```

Comment:
I think you need to be explicit about rounding here, IIRC this is something where ruby and javascript (opal) can produce different results

Reply:
Is there a safe way of doing this? From my reading of the (very brief) Opal docs [[https://opalrb.com/docs/guides/v1.7.2/unsupported_features](https://opalrb.com/docs/guides/v1.7.2/unsupported_features)] this will be integer division in Opal as the divisor is an integer. The terrain costs at 20, 40, 80 and 120, so I believe this should be fine.

Reply:
> Is there a safe way of doing this?

There are a few places where the division uses a float for the divisor and then `ceil` to round up:

https://github.com/tobymao/18xx/blob/a35337642a6a6f8ea36e046c21e4b9dfd48b38ad/lib/engine/company.rb#L28

If you call `.to_i` on the integer, Ruby (and I presume Opal) will round that down.

> From my reading of the (very brief) Opal docs [https://opalrb.com/docs/guides/v1.7.2/unsupported_features] this will be integer division in Opal as the divisor is an integer.

> > Integer / Float difference
> >
> > In Opal, both integers and floats belong to same class `Number` (using JavaScript native numbers). So `1 / 4` is `0.25` (not `0`) and `4.0 / 2` is `2` (not `2.0`).

So if the divisor is the integer `2`, and the numerator is _odd_, you'd end up with a floating point in Opal, but an integer in Ruby. Note the `1/4` example.

> The terrain costs are 20, 40, 80 and 120, so I believe this should be fine.

Given that, I think everything should be fine without further changes.